File: ocaml/bin/p0027.ml
(* Project Euler: Problem 27 *)

(*
  n^2 + an + b,  where abs(a) < 1000, abs(b) < 1000 and n=0...

  Find the product of the coefficients, 'a' and 'b', for the quadratic expression that
  produces the maximum number of x_primes for consecutive values of 'n', starting with 'n'=0.
 *)

(*
  when 'n' = 0:
    'b' must be a prime number. so, 1 < 'b' < 1000.
  when 'n' = 1:
    '1 + a + b' must be a prime number. Assume this prime number is 'x', 'a' = 'x' - 'b' - 1.
    Therefore, abs('x' - b - 1) < 1000 and 1 < 'b' < 1000, hence 0 < 'x' < 2000.
  when 'n' is odd number:
    'n^2 + b' is a even number. so 'a' must be a odd number.
  the value of 'b' cannot be 2:
    bacause value of the expression becomes even when 'n' is an even number. it's not a prime number.
 *)

open Core

let count_consecutive a b =
  let rec loop n =
    if Bool.equal (Euler.Math.mr_isprime (n * n + a * n + b)) false then n else loop (succ n)
  in
  loop 0

let solve () =
  let x_primes = Euler.Eratosthenes.to_list (Euler.Eratosthenes.generate 2_000) in    (* candidates of 'x' *)
  let b_primes = List.tl_exn (List.filter x_primes ~f:(fun n -> n < 1000)) in    (* candidates of 'b' *)
  let rec aux l max_len max_tpl =
    match l with
    | [] -> max_len, max_tpl
    | b :: tl ->
       let rec aux' a_lst max_len max_tpl =
         match a_lst with
         | [] -> max_len, max_tpl
         | a :: tl ->
            let len = count_consecutive a b in
            if len > max_len then
              aux' tl len (a, b)
            else
              aux' tl max_len max_tpl
       in
       let new_len, new_tpl =
         aux' (List.map ~f:(fun x -> x - b - 1) @@ List.filter x_primes ~f:(fun x -> (abs (x - b - 1)) < 1000)) max_len max_tpl in
       aux tl new_len new_tpl
  in
  aux b_primes 0 (0, 0)

let exec () =
  let len, (a, b) = solve () in
  sprintf "%d (a=%d, b=%d, length=%d)" (a * b) a b len

let () = Euler.Task.run exec

File: rust/src/bin/p0027.rs
// Project Euler: Problem 27

/*
  n^2 + an + b,  where abs(a) < 1000 and abs(b) < 1000

  when 'n' = 0:
    '0 + 0 + b' = 'b' must be a prime number. so, 2 < 'b' < 1000.
    ('b' must not be 2 since value of the expression becomes even when 'n' is an even number)
  when 'n' = 1:
    '1 + a + b' must be a prime number. write this prime number is 'x', then 'a' = 'x' - 'b' - 1.
    abs('x' - b - 1) < 1000 and 2 < 'b' < 1000 ===> 0 < 'x' < 2000
  when 'n' is a odd number:
    'n^2 + b' is a even number. so 'a' must be a odd number.
*/

use euler::math::primes;

euler::run_solver!(27);

fn solve() -> String {
    compute().to_string()
}

fn compute() -> i64 {
    let p_lst: Vec<i64> = primes::primes(1, 2_000)
        .into_iter()
        .map(|x| i64::try_from(x).unwrap())
        .collect();

    let (a, b) = p_lst
        .iter()
        .skip(1)
        .filter(|&&p| p < 1_000)
        .flat_map(|&b| {
            p_lst.iter().filter_map(move |&p| {
                let a = p - b - 1;
                if a < 1_000 { Some((a, b)) } else { None }
            })
        })
        .max_by_key(|&(a, b)| count_consec_times(a, b))
        .unwrap();

    a * b
}

fn count_consec_times(a: i64, b: i64) -> u64 {
    let mut n = 0_i64;

    loop {
        let p = n * n + a * n + b;
        if p > 0 && primes::is_prime(p as u64) {
            n += 1;
        } else {
            break;
        }
    }

    u64::try_from(n).unwrap()
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0027() {
        assert_eq!(compute(), -59231);
    }
}
